Sources/Leaf/LeafEncoder.swift

```diff
@@ -1,231 +1,219 @@
+            /// Use a subencoder to create a nested container so the coding paths are correctly maintained.
+            /// Save the subcontainer in our data so it can be resolved later before returning it.
+            .init(self.insert(EncoderImpl(subdecoding: self.encoder, withKey: key).rawContainer(keyedBy: NK.self), forKey: key, as: EncoderKeyedContainerImpl<NK>.self))
+        }
 
-    func encode<T>(_ value: T) throws where T: Encodable {
-        defer { self.count += 1 }
-        if let leafRepresentable = value as? LeafDataRepresentable {
-            self.array.append(.data(leafRepresentable.leafData))
-        } else {
-            let encoder = _Encoder(codingPath: codingPath)
-            try value.encode(to: encoder)
-            if let safeData = encoder.container!.data {
-                self.array.append(safeData)
-            }
+        /// See ``KeyedEncodingContainerProtocol/nestedUnkeyedContainer(forKey:)``.
+        func nestedUnkeyedContainer(forKey key: Key) -> UnkeyedEncodingContainer {
+            self.insert((EncoderImpl(subdecoding: self.encoder, withKey: key).unkeyedContainer() as! EncoderUnkeyedContainerImpl), forKey: key)
         }
-    }
 
-    /// See UnkeyedEncodingContainer.nestedContainer
-    func nestedContainer<NestedKey>(keyedBy keyType: NestedKey.Type) -> KeyedEncodingContainer<NestedKey>
-        where NestedKey: CodingKey
-    {
-        defer { self.count += 1 }
-        let container = KeyedContainer<NestedKey>(codingPath: self.codingPath)
-        self.array.append(.container(container))
-        return .init(container)
-    }
+        /// A super encoder is, in fact, just a subdecoder with delusions of grandeur and some rather haughty
+        /// pretensions. (It's mostly Codable's fault anyway.)
+        func superEncoder() -> Encoder {
+            self.insert(EncoderImpl(subdecoding: self.encoder, withKey: GenericCodingKey(stringValue: "super")), forKey: GenericCodingKey(stringValue: "super"))
+        }
 
-    /// See UnkeyedEncodingContainer.nestedUnkeyedContainer
-    func nestedUnkeyedContainer() -> UnkeyedEncodingContainer {
-        defer { self.count += 1 }
-        let container = UnkeyedContainer(codingPath: self.codingPath)
-        self.array.append(.container(container))
-        return container
-    }
+        /// See ``KeyedEncodingContainerProtocol/superEncoder(forKey:)``.
+        func superEncoder(forKey key: Key) -> Encoder { self.insert(EncoderImpl(subdecoding: self.encoder, withKey: key), forKey: key) }
+    }
+    
+    private final class EncoderUnkeyedContainerImpl: UnkeyedEncodingContainer, LeafEncodingResolvable {
+        let encoder: EncoderImpl
+        var data: [LeafEncodingResolvable] = []
+        var nextCodingKey: CodingKey { GenericCodingKey(intValue: self.count) }
+        var resolvedData: LeafData? { let compact = data.compactMap(\.resolvedData); return compact.isEmpty ? nil : .array(compact) }
```

Comment:
Was this intentional? It now means that in templates I need to do:

```
#if(myArray):
  #if(count(myArray) > 0):
    …
  #endif
#endif
```

Which seems to be a regression of how templates are written.